design-proposals/klip-47-non-null-columns-and-component-types.md

```diff
@@ -0,0 +1,183 @@
+https://github.com/confluentinc/ksql/issues/4436
+
+The SQL standard allows a column to have a `NOT NULL` constraint. `NOT NULL` indicates that
+`NULL` is not an accepted value for that column. This constraint prevents records with a `NULL` column value from being written into the table. 
+When a user reads from the table, its guaranteed that every table record will have a value in the `NOT NULL` column.
+
+Currently, all columns in KSQL are implicitly nullable. This KLIP would introduce `NOT NULL` syntax so users can apply it as a constraint to ksqlDB columns.
+Columns that don't have `NOT NULL` constraint default to being nullable, which is the current behavior. When a `NOT NULL` constraint is applied to a column, 
+records with null column values should not be written into constrained stream/table via `CSAS/CTAS/Insert Into` queries or `Insert Into Values` statements.
+When a user reads from a ksqlDB stream/table with a `NOT NULL` column, they should not read any records that violate the `NOT NULL` constraint.
+
+Being able to specify `NOT NULL` as part of a stream/table schema would also help improve the integration with Schema Registry.
+All ksqlDB columns must have an internal SQL type associated with it. When ksqlDB converts its internal SQL types to a Connect Schema, the schema is always optional (nullable) because there's no 
+existing mechanism/syntax to indicate the type is not nullable. This leads to some undesirable behavior when users use Schema Registry.
+For example, if a map key is nullable, Schema Registry will interpret the map schema as a list of key/value map entries. Since ksqlDB map keys
+are always nullable, ksqlDB will take schemas with maps and write them to Schema Registry as a list of key/value map entries, which is a confusing behavior for users.
+Having map keys that are `NOT NULL` would allow ksqlDB maps to be represented properly in Schema Registry.
+* https://github.com/confluentinc/ksql/issues/6049
+* https://github.com/confluentinc/ksql/issues/2619
+
+For the rest of the KLIP, I'll refer to any stream/table with a `NOT NULL` column as a `NOT NULL` stream/table.
+
+### Data Ownership
+Since ksqlDB relies on Kafka to store the data it reads/writes, the owner of the data in a topic isn't solely the ksqlDB service.
+Producers outside ksqlDB are able to write into the topic. This means that the underlying topic for a `NOT NULL` stream/table can still
+have null column records in it. There's no way around this as ksqlDB can't prevent outside producers from writing into the Kafka topics.
+
+With that in mind, the `NOT NULL` constraint can be enforced within the ksqlDB service. When reading records from a source that has a `NOT NULL` 
+column constraint, invalid records will be dropped, and an error sent to the processing log. When attempting to write an invalid
+record to a sink, the record would also be dropped, and an error sent to the processing log. A `INSERT INTO VALUES` statement with
```

Comment:
What we would include in that processing log entry?

Reply:
> When attempting to write an invalidrecord to a sink

This seems only to apply to `INSERT INTO AS SELECT` (might be worth to clarify).